CMakeLists.txt

```diff
@@ -190,36 +190,8 @@ if(ZFP_WITH_CUDA)
 endif()
 
 if(ZFP_WITH_HIP)
-  if (DEFINED ENV{HIP_PATH} AND NOT $ENV{HIP_PATH} STREQUAL "")
-    set(HIP_PATH $ENV{HIP_PATH} CACHE PATH "Path to which HIP has been installed")
-  endif()
-  set(CMAKE_MODULE_PATH "${HIP_PATH}/cmake" ${CMAKE_MODULE_PATH})
-  find_package(HIP REQUIRED)
-  if(HIP_FOUND)
-    message(STATUS "Found HIP(${HIP_VERSION}) in ${HIP_PATH}")
-  else()
-    message(FATAL_ERROR "Could not find HIP. Ensure that HIP is either installed in /opt/rocm/hip or the variable HIP_PATH is set to point to the right location.")
-  endif()
-  find_package(hipcub REQUIRED)
-  if(hipcub_FOUND)
-    message(STATUS "Found hipCUB.")
-  else()
-    message(FETAL_ERROR "Cound not find hipCUB.")
-  endif()
-  set(CMAKE_CXX_STANDARD 11)
-endif()
-
-if(ZFP_WITH_HIP)
-  if (DEFINED ENV{HIP_PATH} AND NOT $ENV{HIP_PATH} STREQUAL "")
-    set(HIP_PATH $ENV{HIP_PATH} CACHE PATH "Path to which HIP has been installed")
-  endif()
-  set(CMAKE_MODULE_PATH "${HIP_PATH}/cmake" ${CMAKE_MODULE_PATH})
-  find_package(HIP REQUIRED)
-  if(HIP_FOUND)
-    message(STATUS "Found HIP(${HIP_VERSION}) in ${HIP_PATH}")
-  else()
-    message(FATAL_ERROR "Could not find HIP. Ensure that HIP is either installed in /opt/rocm/hip or the variable HIP_PATH is set to point to the right location.")
-  endif()
+  enable_language(HIP)
```

Comment:
This is clearly a lot cleaner, but enable_language(HIP) is available only as of CMake 3.21.  zfp currently requires CMake 3.9.

We have gone to great lengths to make zfp easy to build, with as few requirements and dependencies as possible, and with conformance to old C89 and C++98.  We need to consider the implications of jumping from CMake 3.9 to 3.21 and from C++98 to C++14 just to support HIP.

Reply:
@lindstro Due to the changes in ROCM 5.1.0, we can no longer build ZFP with the existing CMake build system and C++98. This is because ROCM 5.1.0 starts to rely on features in C++14 (e.g., using `std::conditional_t` type). Also, when using the old way (currently used in ZFP) to handle HIP code in CMake (i.e., using `set_source_files_properties(${zfp_source} PROPERTIES HIP_SOURCE_PROPERTY_FORMAT 1)`), the `-std=c++14` flag cannot be correctly passed to the `hipcc` compiler, since CMake cannot correctly recognize the source code language. To let CMake know our source code is written in HIP, we have to `enable_language(HIP)` and `set_source_files_properties(${HIPZFP_SOURCE} PROPERTIES LANGUAGE HIP)`. So, it seems that using `enable_language(HIP)` is our only option to let ZFP build with the latest ROCM. Perhaps it is better to conditional require a newer CMake and C++ versions? For example, we can conditional require CMake 3.21+ and C++14 (or newer) only if users want to build ZFP with HIP.

Reply:
Is C++14 needed even if you don't use such features, e.g., do the headers not compile with pre C++14 compilers?

I agree that we should conditionally require CMake 3.21 and C++14 only when HIP support is enabled.  Can you update the PR to do so?

We should also be better about testing builds with the oldest version of CMake required to catch the use of unsupported CMake features.  That's on us.

Reply:
Those C++14 features are not used in the implementation of ZFP. They are actually used in the HIP runtime system (starting from v5.1.0). So, C++14 is required if we include the HIP runtime header in ZFP. We have no control on whether or not to use those C++14 features. If we compile with pre C++14 compilers, the compiler will complain that the definition of `std::conditional_t` type cannot be found. I will update the PR to add conditional requirements in CMake.